contracts/modules/TokenExchanger.sol

```diff
@@ -33,139 +35,238 @@ contract TokenExchanger is OnlyOwnerModule {
+        verifyExchangeAdapters(_routes);
+        // Approve source amount if required
+        uint previousAllowance = approveToken(_wallet, _srcToken, _maxSrcAmount);
+        // Perform trade and emit event
+        doBuy(
+            _wallet,
+            _srcToken,
+            _destToken,
+            _maxSrcAmount,
+            _destAmount,
+            _expectedDestAmount,
+            _routes,
+            _mintPrice);
+        // Restore the previous allowance if needed (paraswap.buy() may not have used exactly the additional allowance granted to it)
+        restoreAllowance(_wallet, _srcToken, previousAllowance);
+    }
+
+    // Internal & Private Methods
+
+    function verifyExchangeAdapters(IAugustusSwapper.Path[] calldata _path) internal view {
+        for (uint i = 0; i < _path.length; i++) {
+            for (uint j = 0; j < _path[i].routes.length; j++) {
+                require(authorisedExchanges[_path[i].routes[j].exchange], "TE: Unauthorised Exchange");
+            }
         }
-        if (_srcToken == ETH_TOKEN_ADDRESS) {
-            uint256 srcTradable = _srcAmount.sub(fee);
-            methodData = abi.encodeWithSignature(
-                "trade(address,uint256,address,address,uint256,uint256,address)",
-                _srcToken,
-                srcTradable,
-                _destToken,
-                _wallet,
-                _maxDestAmount,
-                _minConversionRate,
-                feeCollector
-                );
-            invokeWallet(_wallet, kyber, srcTradable, methodData);
-        } else {
-            // approve kyber on erc20
-            methodData = abi.encodeWithSignature("approve(address,uint256)", kyber, _srcAmount);
-            invokeWallet(_wallet, _srcToken, 0, methodData);
-            // transfer erc20
-            methodData = abi.encodeWithSignature(
-                "trade(address,uint256,address,address,uint256,uint256,address)",
-                _srcToken,
-                _srcAmount,
-                _destToken,
-                _wallet,
-                _maxDestAmount,
-                _minConversionRate,
-                feeCollector
-                );
-            invokeWallet(_wallet, kyber, 0, methodData);
+    }
+
+    function verifyExchangeAdapters(IAugustusSwapper.BuyRoute[] calldata _routes) internal view {
+        for (uint j = 0; j < _routes.length; j++) {
+            require(authorisedExchanges[_routes[j].exchange], "TE: Unauthorised Exchange");
         }
+    }
 
-        if (fee > 0) {
-            invokeWallet(_wallet, feeCollector, fee, "");
+    function approveToken(address _wallet, address _token, uint _amount) internal returns (uint256 _existingAllowance) {
+        // TODO: Use a "safe approve" logic similar to the one implemented below in other modules
+        if (_token != ETH_TOKEN_ADDRESS) {
+            _existingAllowance = ERC20(_token).allowance(_wallet, paraswapProxy);
+            if (_existingAllowance < uint256(-1)) {
```

Comment:
Why is this condition necessary?

Reply:
If the user has given an unlimited allowance to Paraswap (via WalletConnect), we should not change the allowance. That's not only to save gas in this particular transaction, but also because many tokens (DAI, USDT, etc) optimise `transferFrom` when the allowance is `uint(-1)` so as to save gas and I think we should not break this optimisation when the user opts for it.

Reply:
Also, without this condition `SafeMath.add(_existingAllowance, _amount)` would revert when  `_existingAllowance == uint(-1)`.